src/domains.rs

```diff
@@ -0,0 +1,199 @@
+
+                if !userinfo_allowed && require_host {
+                    // We don't have to look further
+                    break;
+                }
+
+                false
+            }
+            ':' => {
+                // Could be in userinfo, or we're getting a port now.
+                if !userinfo_allowed && !port_allowed {
+                    break;
+                }
+
+                // Don't advance the last dot when we get to port numbers
+                maybe_last_dot = last_dot;
+
+                false
+            }
+            '@' => {
+                if !userinfo_allowed {
+                    // We already had userinfo, can't have another `@` in a valid authority.
+                    return (None, None);
+                }
+
+                // Sike! Everything before this has been userinfo, so let's reset our
+                // opinions about all the host bits.
+                userinfo_allowed = false;
+
+                maybe_last_dot = None;
+                last_dot = None;
+                dot_allowed = false;
+                hyphen_allowed = false;
+                all_numeric = true;
+                maybe_host = true;
+                host_ended = false;
+
+                false
+            }
+            '/' => {
+                if !require_host {
+                    // For schemes where we allow anything, we want to stop at delimiter characters
+                    // except if we get a slash closing the URL, which happened here.
+                    end = Some(i);
+                }
+                break;
+            }
+            _ => {
+                // Anything else, this might be the end of the authority (can be empty).
+                // Now let the rest of the code handle checking whether the end of the URL is
+                // valid.
+                break;
+            }
+        };
+
+        if can_be_last {
+            end = Some(i + c.len_utf8());
+        }
+    }
+
+    if require_host {
+        if maybe_host {
+            // Can't have just a number without dots as the authority
+            if all_numeric && last_dot.is_none() && end != Some(0) {
+                return (None, None);
+            }
+
+            // If we have something that is not just numeric (not an IP address),
+            // check that the TLD looks reasonable. This is to avoid linking things like
+            // `abc@v1.1`.
+            if !all_numeric {
+                if let Some(last_dot) = last_dot {
+                    if !valid_tld(&s[last_dot + 1..]) {
+                        return (None, None);
+                    }
+                }
+            }
+
+            return (end, last_dot);
+        } else {
+            return (None, None);
+        }
+    } else {
+        return (end, last_dot);
+    }
+}
+
+fn valid_tld(tld: &str) -> bool {
+    tld.chars()
+        .take_while(|c| c.is_ascii_alphabetic())
+        .take(2)
+        .count()
+        >= 2
```

Comment:
Confused by that. How can this be bigger than 2 if we `take` 2 elements from the iterator? Perhaps you also want to add some unit tests.

Reply:
Good eye. You're right, it can't, using `>= 2` and `== 2` have exactly the same result here. But I still prefer to write it this way because it communicates the intent of the code more clearly, which is that we need *at least* 2 letters (not exactly 2 letters). I also like it because you could take away the `.take(2)` optimization and it would still be correct, whereas the `== 2` variant would be incorrect without the take.